# One way anova

Researchers took 20 cars of the same to take part in a study. These cars are randomly doped with one of the four-engine oils and allowed to run freely for 100 kilometers each. At the end of the journey, the performance of each of the cars is noted.

In [1]:
performance1 = [89, 89, 88, 78, 79]
performance2 = [93, 92, 94, 89, 88]
performance3 = [89, 88, 89, 93, 90]
performance4 = [81, 78, 81, 92, 82]

In [2]:
from scipy.stats import f_oneway
 
# Performance when each of the engine 
# oil is applied
performance1 = [89, 89, 88, 78, 79]
performance2 = [93, 92, 94, 89, 88]
performance3 = [89, 88, 89, 93, 90]
performance4 = [81, 78, 81, 92, 82]
 
# Conduct the one-way ANOVA
f_oneway(performance1, performance2, performance3, performance4)

F_onewayResult(statistic=4.625000000000002, pvalue=0.016336459839780215)

### The F statistic and p-value turn out to be equal to 4.625 and 0.016336498 respectively. Since the p-value is less than 0.05 hence we would reject the null hypothesis.

This implies that we have sufficient proof to say that there exists a difference in the performance among four different engine oils. 

# Two way anova

Let us consider an example in which scientists need to know whether plant growth is affected by fertilizers and watering frequency. They planted exactly 30 plants and allowed them to grow for six months under different conditions for fertilizers and watering frequency. After exactly six months, they recorded the heights of each plant centimeters.

In [3]:
import numpy as np 
import pandas as pd

In [4]:
# Importing libraries 
import numpy as np 
import pandas as pd 
  
# Create a dataframe 
dataframe = pd.DataFrame({'Fertilizer': np.repeat(['daily', 'weekly'], 15), 
                          'Watering': np.repeat(['daily', 'weekly'], 15), 
                          'height': [14, 16, 15, 15, 16, 13, 12, 11, 14,  
                                     15, 16, 16, 17, 18, 14, 13, 14, 14,  
                                     14, 15, 16, 16, 17, 18, 14, 13, 14,  
                                     14, 14, 15]}) 

In [10]:
import statsmodels.api as sm 
from statsmodels.formula.api import ols 

In [11]:
# Performing two-way ANOVA
model = ols('height ~ C(Fertilizer) + C(Watering) + \
             C(Fertilizer):C(Watering)', 
            data=dataframe).fit() 
result = sm.stats.anova_lm(model, type=2)

# Print the result
print(result)

                             df     sum_sq   mean_sq         F    PR(>F)
C(Fertilizer)               1.0   0.033333  0.033333  0.012069  0.913305
C(Watering)                 1.0   1.027463  1.027463  0.372012  0.546828
C(Fertilizer):C(Watering)   1.0   0.133582  0.133582  0.048366  0.827527
Residual                   28.0  77.333333  2.761905       NaN       NaN


## Following are the p-values for each of the factors in the output:

### The fertilizer p-value is equal to 0.913305

### The Watering p-value is equal to 0.990865

### The Fertilizer * Watering: p-value is equal to 0.904053

The p-values for water and sun turn out to be less than 0.05 which implies that the means of both the factors possess a statistically significant effect on plant height. The p-value for the interaction effect (0.904053) is greater than 0.05 which depicts that there is no significant interaction effect between fertilizer frequency and watering frequency.